launch_ros/launch_ros/actions/node.py

```diff
@@ -138,43 +114,23 @@ def __init__(
         :param: remappings ordered list of 'to' and 'from' string pairs to be
             passed to the node as ROS remapping rules
         :param: arguments list of extra arguments for the node
+        :param additional_env: Dictionary of environment variables to be added. If env was
+            None, they are added to the current environment. If not, env is updated with
+            additional_env.
         """
-        if package is not None:
-            cmd = [ExecutableInPackage(package=package, executable=executable)]
-        else:
-            cmd = [executable]
-        cmd += [] if arguments is None else arguments
-        # Reserve space for ros specific arguments.
-        # The substitutions will get expanded when the action is executed.
-        cmd += ['--ros-args']  # Prepend ros specific arguments with --ros-args flag
-        if name is not None:
-            cmd += ['-r', LocalSubstitution(
-                "ros_specific_arguments['name']", description='node name')]
-        if parameters is not None:
-            ensure_argument_type(parameters, (list), 'parameters', 'Node')
-            # All elements in the list are paths to files with parameters (or substitutions that
-            # evaluate to paths), or dictionaries of parameters (fields can be substitutions).
-            normalized_params = normalize_parameters(parameters)
-        # Forward 'exec_name' as to ExecuteProcess constructor
-        kwargs['name'] = exec_name
-        super().__init__(cmd=cmd, **kwargs)
-        self.__package = package
-        self.__node_executable = executable
-        self.__node_name = name
-        self.__node_namespace = namespace
-        self.__parameters = [] if parameters is None else normalized_params
-        self.__remappings = [] if remappings is None else list(normalize_remap_rules(remappings))
-        self.__arguments = arguments
-
-        self.__expanded_node_name = self.UNSPECIFIED_NODE_NAME
-        self.__expanded_node_namespace = self.UNSPECIFIED_NODE_NAMESPACE
-        self.__expanded_parameter_arguments = None  # type: Optional[List[Tuple[Text, bool]]]
-        self.__final_node_name = None  # type: Optional[Text]
-        self.__expanded_remappings = None  # type: Optional[List[Tuple[Text, Text]]]
+        self.__node_desc = NodeDescription(node_name=name, node_namespace=namespace,
+                                           parameters=parameters, remappings=remappings)
+        ros_exec_kwargs = {'additional_env': additional_env, 'arguments': arguments}
+        self.__ros_exec = RosExecutable(package=package, executable=executable,
+                                        nodes=[self.__node_desc], **ros_exec_kwargs)
+        super().__init__(process_description=self.__ros_exec, **kwargs)
 
-        self.__substitutions_performed = False
+    def prepare(self, context: LaunchContext):
+        self.__node_desc.prepare(context, self.__ros_exec, self)
```

Comment:
@roger-strain why preparing the node description here? Won't `ExecuteLocal.prepare()` invoke `RosExecutable.prepare()`, which in turn will invoke `NodeDescription.prepare()`? That'd make sense to me.

Reply:
On reflection, I believe you're correct. I think this was probably added earlier in the development cycle, before the `RosExecutable` portion came in.